sdk/core/azure-core-amqp/src/amqp/connection.cpp

```diff
@@ -0,0 +1,401 @@
+  {
+    m_impl->SetIdleTimeout(timeout);
+  }
+  void Connection::SetRemoteIdleTimeoutEmptyFrameSendRatio(double idleTimeoutEmptyFrameSendRatio)
+  {
+    return m_impl->SetRemoteIdleTimeoutEmptyFrameSendRatio(idleTimeoutEmptyFrameSendRatio);
+  }
+
+  void Connection::SetProperties(Azure::Core::Amqp::Models::AmqpValue properties)
+  {
+    m_impl->SetProperties(properties);
+  }
+  Azure::Core::Amqp::Models::AmqpValue Connection::GetProperties() const
+  {
+    return m_impl->GetProperties();
+  }
+}}}} // namespace Azure::Core::Amqp::_internal
+
+namespace Azure { namespace Core { namespace Amqp { namespace _detail {
+  namespace {
+    void EnsureGlobalStateInitialized()
+    {
+      // Force the global instance to exist. This is required to ensure that uAMQP and
+      // azure-c-shared-utility is
+      auto globalInstance
+          = Azure::Core::Amqp::Common::_detail::GlobalStateHolder::GlobalStateInstance();
+      (void)globalInstance;
+    }
+  } // namespace
+
+  // Create a connection with an existing networking Transport.
+  ConnectionImpl::ConnectionImpl(
+      std::shared_ptr<Network::_detail::TransportImpl> transport,
+      _internal::ConnectionOptions const& options,
+      _internal::ConnectionEvents* eventHandler)
+      : m_hostName{options.HostName}, m_options{options}, m_eventHandler{eventHandler}
+  {
+    if (options.SaslCredentials)
+    {
+      throw std::runtime_error("Sasl Credentials should not be provided with a transport.");
+    }
+    EnsureGlobalStateInitialized();
+    m_transport = transport;
+  }
+
+  // Create a connection with a request URI and options.
+  ConnectionImpl::ConnectionImpl(
+      std::string const& requestUri,
+      _internal::ConnectionOptions const& options,
+      _internal::ConnectionEvents* eventHandler)
+      : m_options{options}, m_eventHandler{eventHandler}
+  {
+    EnsureGlobalStateInitialized();
+
+    if (options.SaslCredentials)
+    {
+      throw std::runtime_error("Sasl Credentials should not be provided with a request URI.");
+    }
+    Azure::Core::Url requestUrl(requestUri);
+    std::shared_ptr<Network::_internal::Transport> requestTransport;
+    if (requestUrl.GetScheme() == "amqp")
+    {
+      Log::Write(Logger::Level::Informational, "Creating socket connection transport.");
+      Azure::Core::Amqp::Network::_internal::SocketTransport transport{
+          requestUrl.GetHost(),
+          requestUrl.GetPort() ? requestUrl.GetPort() : static_cast<std::uint16_t>(AmqpPort)};
+      m_transport = transport.GetImpl();
+    }
+    else if (requestUrl.GetScheme() == "amqps")
+    {
+      Log::Write(Logger::Level::Informational, "Creating TLS socket connection transport.");
+      Azure::Core::Amqp::Network::_internal::TlsTransport transport{
+          requestUrl.GetHost(),
+          requestUrl.GetPort() ? requestUrl.GetPort() : static_cast<std::uint16_t>(AmqpPort)};
+      m_transport = transport.GetImpl();
+    }
```

Comment:
nit: rather than calling `GetPort()` twice, cache in a local

It's interesting you are relying on `GetPort()` returning 0 to do a bool test on the return value :)

Reply:
`Azure::Core::Url::GetPort` is defined as returning 0 if no port is specified in the URL, this code depends on this well defined behavior (it's covered in a @note on Azure::Core::Url). 

Also, when retrieving an int value from a struct, any performance benefit from locally caching the value is almost non-existent (you're still fetching a value from memory, it's just a question of fetching memory from a location indexed from the rbp register or the rcx register).

Reply:
> Also, when retrieving an int value from a struct

That's fair, but you are relying on an implementation detail, which could change (though very unlikely).